### PR TITLE
bazel: rewrite MD5 .note.gnu.build-id with truncated git SHA1.

### PR DESCRIPTION
**Pull request by htuch**
This is a workaround for
https://github.com/bazelbuild/bazel/issues/2805.

Small reorganization of version_generated.cc rules, since we rely on them
here and this can also assist with the Google import of Envoy and its
versioning scheme.

Also added docs on build types and how to do release builds.

### CODE REVIEW

File: bazel/README.md [hunk 1]
```diff
@@ -71,13 +71,42 @@ tools/bazel-test-gdb //test/common/http:async_client_impl_test
 
 # Additional Envoy build and test options
 
+In general, there are 3 [compilation
+modes](https://bazel.build/versions/master/docs/bazel-user-manual.html#flag--compilation_mode)
+that Bazel supports:
+
+* `fastbuild`: `-O0 -DDEBUG`, aimed at developer speed (default).
```
Comment (mattklein123) [inline bazel/README.md]: Per other PR, let's just kill -DDEBUG after checking we don't use it?

File: bazel/README.md [hunk 2]
```diff
@@ -71,13 +71,42 @@ tools/bazel-test-gdb //test/common/http:async_client_impl_test
...
+* `opt`: `-O2`, for production builds and performance benchmarking.
```
Comment (mattklein123) [inline bazel/README.md]: This is related to https://github.com/lyft/envoy/pull/720, but I definitely at least want the ability to set -ggdb3 on opt builds. (IMO there is no reason to not generate debugging info and have easy access when looking at core dumps, but to each there on). If we have an option builders can choose.

File: tools/git_sha_rewriter.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+# a 16 byte truncated git SHA1, rather than a complete 20 byte git SHA1.
+# This is a workaround to https://github.com/bazelbuild/bazel/issues/2805.
+
+import binascii
+import re
+import subprocess as sp
+import sys
+
+# This is what the part of .note.gnu.build-id prior to the MD5 hash looks like.
+EXPECTED_BUILD_ID_NOTE_PREFIX = [
```
Comment (mattklein123) [inline tools/git_sha_rewriter.py]: Your bazel hacking skills continue to astound me. 😉

File: bazel/README.md [hunk 4]
```diff
@@ -71,13 +71,42 @@ tools/bazel-test-gdb //test/common/http:async_client_impl_test
...
+
+# Release builds
+
+Release builds should be built in `opt` mode, processed with `strip` and have a
+`.note.gnu.build-id` section with the Git SHA1 at which the build took place.
+They should also ignore any local `.bazelrc` for reproducibility. This can be
+achieved with:
+
+```
+bazel --bazelrc=/dev/null build -c opt //source/exe:envoy-static.stripped.stamped
```
Comment (mattklein123) [inline bazel/README.md]: If I read correctly below, //source/exe:envoy-static.stamped is valid, right? (If I want a stamped, but not stripped build)?

File: bazel/README.md (hunk 2 above)
Comment (htuch) [inline bazel/README.md]: Let me add this there and here (I should have split out this doco into a separate PR, but oh well).

File: bazel/README.md (hunk 4 above)
Comment (htuch) [inline bazel/README.md]: Yep.

File: bazel/README.md (hunk 1 above)
Comment (htuch) [inline bazel/README.md]: Done.

File: bazel/README.md (hunk 2 above)
Comment (htuch) [inline bazel/README.md]: I've added `--define debug_symbols=yes` support. I actually think it probably would be just as valid to have just suggested that the invoker provider `--copt=-ggdb3` as an extra `bazel build` arg, but this works too.

**Comment by mattklein123** [comment]
@htuch sorry needs master merge again
